sdk/docs/core/README.md

```diff
@@ -177,6 +177,138 @@ Note however that cancellation is performed as a best effort; it is not guarante
+
+   az_mqtt5_rpc_client_init(
+      &rpc_client,
+      &rpc_client_codec,
+      &mqtt_connection,
+      /*...*/);
+
+  az_mqtt_connection_begin_open(&mqtt_connection); 
+
+  while(!done) { sleep(1000); }
+}
+
+// The mqtt_callback function is guaranteed to be thread-safe with respect to the connection object.
+// Note that if the callback is shared by multiple connections, it must be designed to be reentrant.
+az_result mqtt_callback(az_mqtt_connection* connection, az_event event) 
+{ 
+  // No blocking calls allowed within the callback.  
+  // Long running (CPU-bound) operations will stall the network stack. 
+  switch (event.type) 
+  { 
+    case AZ_MQTT_EVENT_CONNECT_RSP:
+      az_mqtt5_rpc_client_invoke_req_event_data command_data = { /* ... */ };
+      az_mqtt5_rpc_client_begin_invoke_request(&rpc_client, &rpc_context, &command_data);
+      break;
+
+    case AZ_MQTT5_EVENT_RPC_CLIENT_RSP: 
+      // The response object lifetime is guaranteed for the duration of this callback. 
+      recv_data = (az_mqtt5_rpc_client_rsp_event_data*)event.data;
+      break;
+
+    // case ... […] 
+    
+    default: 
+      // unhandled event 
+      panic();
+  } 
+}
+```
+
+#### Event Data Lifetime
+
+The main function is used for initial set-up. After that, the main application thread may be reused for other purposes. All operation completions as well as network events will be handled by the single `mqtt_callback` function.
+
+The SDK is designed to avoid memory copy which results in the following ownership and life-time allocation requirements:
+
+- __Sending:__ The application may call APIs from any thread. The application is responsible with event-data allocation and initialization. The lifetime of the event-data associated with the API call will depend on the type of operation. It is safe to deallocate or re-use the event-data object after the operation completed (either successfully or with error). Please see the API documentation for more details. E.g. `az_mqtt5_rpc_client_begin_invoke_request` requires that the `az_mqtt5_rpc_client_invoke_req_event_data` object remains allocated until a `AZ_MQTT5_EVENT_RPC_CLIENT_RSP` is received.
+
+- __Receiving:__ When receiving events, the event-data object lifetime is guaranteed only for the duration of the callback. E.g. Receiving the `AZ_MQTT5_EVENT_RPC_CLIENT_RSP` event guarantees that the event-data is of `az_mqtt5_rpc_client_rsp_event_data` type. The lifetime 
+
+Furthermore, long running operations within the callback will stall the network stack (this is intentional and will trigger TCP flow-control to allow the resource-constrained device more time to process inbound data). It is important to remember:
+
+   1. Never save event-data components such as `az_span`s without copying the referenced data into a memory location owned by the application.
+   1. Never block the callback with blocking I/O, synchronization constructs (mutex, semaphores, etc.) or long CPU-bound operations.
```

Comment:
We currently have a mutex in the server sample mqtt_callback to ensure safe copying of the pending command data